runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -3363,3 +3371,103 @@ J9::ValuePropagation::innerConstrainAcall(TR::Node *node)
 
    return node;
    }
+
+TR_OpaqueClassBlock *
+J9::ValuePropagation::findLikelySubtype(TR_OpaqueClassBlock *klass)
+   {
+   if (!klass || TR::VPConstraint::isSpecialClass((uintptr_t)klass) || comp()->compileRelocatableCode())
+      return NULL;
+
+   int32_t numDims = 0;
+   klass = ((TR_J9VM *)fe())->getBaseComponentClass(klass, numDims);
+
+   // If the returned base component class is still an array, it's a primitive array
+   if (TR::Compiler->cls.isClassArray(comp(), klass))
+      return NULL; // primitive array, should be fixed-type anyway
+
+   if (TR::Compiler->cls.isInterfaceClass(comp(), klass)
+       || TR::Compiler->cls.isAbstractClass(comp(), klass))
+      {
+      TR::ClassTableCriticalSection lock(fe());
+      auto *chTable = comp()->getPersistentInfo()->getPersistentCHTable();
+      klass = chTable->findSingleConcreteSubClass(klass, comp());
+      }
+   // If it's a concrete class and it has been extended, there is no particularly likely type
+   else if (fe()->classHasBeenExtended(klass)
+            || TR::Compiler->vm.isVMInStartupPhase(comp()))
+      {
+      return NULL;
+      }
+
+   while (klass != NULL && numDims > 0)
+      {
+      klass = fe()->getArrayClassFromComponentClass(klass);
+      numDims--;
+      }
+
+   return klass;
+   }
+
+TR_OpaqueClassBlock *
+J9::ValuePropagation::findLikelySubtype(const char *sig, int32_t len, TR_ResolvedMethod *owningMethod)
+   {
+   if (!sig || !owningMethod || comp()->compileRelocatableCode())
+      return NULL;
+
+   TR_OpaqueClassBlock *clazz = fe()->getClassFromSignature(sig, len, owningMethod);
+   if (clazz)
+      return findLikelySubtype(clazz);
+
+   // If InterfaceA is implemented by ConcreteB, "[LInterface" might not exist.
+   // Try to see if "[LConcreteB" exists.
+   if (sig[0] == '[')
+      {
+      int32_t numDims=0;
+      while(sig[numDims] == '[' && numDims < len)
```

Comment:
This should check `numDims < len` before accessing `sig[numDims]`, rather than afterward

By the way, there's a space missing between `while` and `(`

Reply:
Whoops, sorry my bad. I should've caught that one. Fixed in 67eca14